debian/tracker/annotation.go

```diff
@@ -0,0 +1,256 @@
+
+	// e.g.
+	// [stretch] - apache2 2.4.25-3+deb9u10
+	// - libredwg <itp> (bug #595191)
+	// ref. https://salsa.debian.org/security-tracker-team/security-tracker/-/blob/50ca55fb66ec7592f9bc1053a11dbf0bd50ee425/lib/python/sectracker/parsers.py#L120-121
+	pkgVersionRegexp = `(?:\[(?P<release>[a-z]+)\]\s)?-\s(?P<package>[A-Za-z0-9:.+-]+)\s*` +
+		`(?:\s(?P<version>[A-Za-z0-9:.+~-]+)\s*)?(?:\s\((?P<inner>.*)\))?`
+
+	// ref. https://salsa.debian.org/security-tracker-team/security-tracker/-/blob/50ca55fb66ec7592f9bc1053a11dbf0bd50ee425/lib/python/sectracker/parsers.py#L142-143
+	pkgPseudoRegexp = `(?:\[(?P<release>[a-z]+)\]\s)?-\s(?P<package>[A-Za-z0-9:.+-]+)` +
+		`\s+<(?P<kind>[a-z-]+)>\s*(?:\s\((?P<inner>.*)\))?`
+
+	// ref. https://salsa.debian.org/security-tracker-team/security-tracker/-/blob/50ca55fb66ec7592f9bc1053a11dbf0bd50ee425/lib/python/sectracker/parsers.py#L175
+	xrefRegexp = `\{(?P<xref>.*)\}`
+
+	// inner annotations, like (bug #1345; low)
+	// ref. https://salsa.debian.org/security-tracker-team/security-tracker/-/blob/50ca55fb66ec7592f9bc1053a11dbf0bd50ee425/lib/python/sectracker/parsers.py#L85-99
+	severityRegexp = regexp.MustCompile(`(unimportant|low|medium|high)`)
+	bugNoRegexp    = regexp.MustCompile(`bug #(?P<bugno>\d+)`)
+
+	// ref. https://salsa.debian.org/security-tracker-team/security-tracker/-/blob/50ca55fb66ec7592f9bc1053a11dbf0bd50ee425/lib/python/sectracker/parsers.py#L140-141
+	pseudoFreeText = []string{"no-dsa", "not-affected", "end-of-life", "ignored", "postponed"}
+	pseudoStruct   = []string{"unfixed", "removed", "itp", "undetermined"}
+)
+
+type Annotation struct {
+	Original    string   `json:",omitempty"`
+	Line        int      `json:",omitempty"`
+	Type        string   `json:",omitempty"`
+	Release     string   `json:",omitempty"`
+	Package     string   `json:",omitempty"`
+	Kind        string   `json:",omitempty"`
+	Version     string   `json:",omitempty"`
+	Description string   `json:",omitempty"`
+	Bugs        []string `json:",omitempty"`
+
+	// inner annotations
+	Severity string `json:",omitempty"`
+	BugNo    int    `json:",omitempty"`
+}
+
+type annotationParser interface {
+	Match(line string) []string
+	Apply(match []string, ann *Annotation)
+}
+
+type annotationDispatcher struct {
+	parsers []annotationParser
+}
+
+func newAnnotationDispatcher() annotationDispatcher {
+	return annotationDispatcher{
+		parsers: []annotationParser{
+			newFlagAnnotation(),
+			newStringAnnotation(),
+			newPkgVersionAnnotation(),
+			newPkgPseudoAnnotation(),
+			newXrefAnnotation(),
+		},
+	}
+}
+
+func (d annotationDispatcher) parseAnnotation(line string, lineno int) *Annotation {
+	var once sync.Once
+	var ann *Annotation
+
+	for _, p := range d.parsers {
+		match := p.Match(line)
+		if len(match) != 0 {
+			once.Do(func() {
+				ann = &Annotation{
+					Original: strings.TrimSpace(line),
+					Line:     lineno,
+				}
+			})
+			p.Apply(match, ann)
```

Comment:
we can break the loop if we find the match.

Reply:
I'm concerned that a single line contains multiple annotations, so I didn't break the loop. Are you sure multiple annotations are not allowed?